File: src/main/java/com/leetcode/OneZeroEight.java
package com.leetcode;

import com.entity.TreeNode;

/**
 * @author: John
 * @date: 2020-02-19 14:26
 * @description: 108
 **/
public class OneZeroEight {


    public TreeNode sortedArrayToBST(int[] num) {
        if (num.length == 0) {
           return null;
        }
        TreeNode head = helper(num, 0, num.length - 1);
        return head;
    }

    public TreeNode helper(int[] num, int low, int high) {
        if (low > high) {
            return null;
        }
        int mid = low + (high - low) / 2;
        TreeNode node = new TreeNode(num[mid]);
        node.left = helper(num, low, mid - 1);
        node.right = helper(num, mid + 1, high);
        return node;
    }
}

File: leetcodecpp/leetcode/OneZeroEight.cpp
//
// Created by John on 2020/2/19.
//

#include "OneZeroEight.h"
#include "TreeNode.h"
#include <vector>

using namespace std;

class Solution{
public:
    TreeNode* sortedArrayToBST(vector<int> & nums) {
        if (nums.size() == 0) {
            return NULL;
        }
        TreeNode * head = helper(nums, 0, nums.size() - 1);
        return head;
    }

private:
    TreeNode* helper(vector<int> nums, int low, int high) {
        if (low > high) {
            return NULL;
        }
        int mid = low + (high - low) / 2;
        TreeNode *node = new TreeNode(nums[mid]);
        node->left = helper(nums, low, mid - 1);
        node->right = helper(nums, mid + 1, high);
        return node;
    }
};